Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/KaalChakra/Chakras.hs

```diff
@@ -46,42 +47,125 @@ mkRunKaalChakraJobReq chakra LYT.KaalChakraJobData {..} = do
+      pure Complete
+
 runDailyJob ::
   ChakraJobs m r c =>
   Job 'Daily ->
   m ExecutionResult
-runDailyJob Job {id, jobInfo, scheduledAt} = withLogTag ("JobId-" <> id.getId) do
+runDailyJob Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
   logInfo "Running Daily Job"
   let req = mkRunKaalChakraJobReq LYT.Daily jobInfo.jobData
-  void $ Event.kaalChakraEvent kaalChakraHandle req
-  pure $ ReSchedule $ Time.incrementDay scheduledAt
+  runKaalChakraAndResheduleJob req
 
 runWeeklyJob ::
   ChakraJobs m r c =>
   Job 'Weekly ->
   m ExecutionResult
-runWeeklyJob Job {id, jobInfo, scheduledAt} = withLogTag ("JobId-" <> id.getId) do
+runWeeklyJob Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
   logInfo "Running Weekly Job"
   let req = mkRunKaalChakraJobReq LYT.Weekly jobInfo.jobData
-  void $ Event.kaalChakraEvent kaalChakraHandle req
-  pure $ ReSchedule $ Time.incrementWeek scheduledAt
+  runKaalChakraAndResheduleJob req
 
 runQuarterlyJob ::
   ChakraJobs m r c =>
   Job 'Quarterly ->
   m ExecutionResult
-runQuarterlyJob Job {id, jobInfo, scheduledAt} = withLogTag ("JobId-" <> id.getId) do
+runQuarterlyJob Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
   logInfo "Running Quarterly Job"
   let req = mkRunKaalChakraJobReq LYT.Quarterly jobInfo.jobData
-  void $ Event.kaalChakraEvent kaalChakraHandle req
-  pure $ ReSchedule $ Time.incrementQuarter scheduledAt
+  runKaalChakraAndResheduleJob req
 
 runMonthlyJob ::
   ChakraJobs m r c =>
   Job 'Monthly ->
   m ExecutionResult
-runMonthlyJob Job {id, jobInfo, scheduledAt} = withLogTag ("JobId-" <> id.getId) do
+runMonthlyJob Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
   logInfo "Running Monthly Job"
   let req = mkRunKaalChakraJobReq LYT.Monthly jobInfo.jobData
-  void $ Event.kaalChakraEvent kaalChakraHandle req
-  pure $ ReSchedule $ Time.incrementMonth scheduledAt
+  runKaalChakraAndResheduleJob req
+
+runKaalChakraUpdateTagsJob ::
+  ChakraJobs m r c =>
+  LYT.UpdateKaalBasedTagsJobReq ->
+  m ExecutionResult
+runKaalChakraUpdateTagsJob req = do
+  eventResult <- Event.updateUserTagsHandler kaalChakraHandle req
+  now <- getCurrentTime
+  case eventResult.chakraBatchState of
+    LYT.Continue shortDelayTime -> do
+      pure $ ReSchedule $ addUTCTime (fromIntegral shortDelayTime) now
+    LYT.Completed -> do
+      void $ Event.clearEventData req.chakra eventResult.eventId
+      let newScheduleTime = getNextChakraTime now req.chakra
+      kaalChakraHandle.createFetchUserDataJob req newScheduleTime
+      pure Complete
+
+-- for now keeping the rescheduling of ckahra job to be at night 12:00 + some delta based on chakra type
+getNextChakraTime :: UTCTime -> LYT.Chakra -> UTCTime
+getNextChakraTime now chakra = do
+  let endOfDayToday = SWC.incrementPeriod SWCT.Days now
+  -- above function gives us the end of day time for the next day,
+  -- i.e. 00:00 of the next day no matter what the time today you run this function.
```

Comment:
what if we want to schedule job at another time?

Reply:
yeah we discussed, will add support for that later in some PR is required, for now its fine.